src/pybind/mgr/mgr_util.py

```diff
@@ -335,6 +338,116 @@ def get_all_filesystems(self) -> List[str]:
         return fs_list
 
 
+class CephFsEarmarkResolver:
+    """
+    Resolves and manages earmarks for CephFS subvolumes.
+
+    This class provides functionality to fetch and manage earmarks for subvolumes,
+    ensuring consistent handling across different modules (NFS, SMB).
+    """
+    def __init__(self, mgr: Module_T, *, client: Optional[CephfsClient] = None) -> None:
+        self._mgr = mgr
+        self._cephfs_client = client or CephfsClient(mgr)
+
+    def _fetch_data(self, command: Dict[str, str], ctx: str) -> Optional[List[Dict[str, str]]]:
+        """Fetch data using the provided command and return a list of dictionaries."""
+        logger.debug('Fetching %s: %r', ctx, command)
+        ret, data, status = self._mgr.mon_command(command)
+        if ret != 0:
+            logger.error('Failed to fetch %s: %r', ctx, status)
+            return None
+        logger.info('Fetched %s: %r', ctx, data)
+        try:
+            return json.loads(data.strip())
+        except json.JSONDecodeError:
+            logger.error('Failed to parse JSON for %s: %r', ctx, data)
+            return None
+
+    def _fetch_subvolume_groups(self, volume: str) -> List[str]:
+        """Fetch subvolumegroups for a given volume."""
+        cmd = {'prefix': 'fs subvolumegroup ls', 'vol_name': volume}
+        data = self._fetch_data(cmd, "subvolumegroups")
+        return [subvolgrp["name"] for subvolgrp in data] if data else []
+
+    def _fetch_subvolumes(self, volume: str, subvolumegrp: str) -> List[str]:
+        """Fetch subvolumes for a given volume and subvolumegroup."""
+        cmd = {
+            'prefix': 'fs subvolume ls',
+            'vol_name': volume,
+            'group_name': subvolumegrp if subvolumegrp != '_nogroup' else '',
+        }
+        data = self._fetch_data(cmd, "subvolumes")
+        return [subvol["name"] for subvol in data] if data else []
```

Comment:
Don't we pretty much already do this in smb/fs.py?  And that code has caching (that you helped add). I am concerned that we're duplicating stuff here. I'd rather lift stuff out of fs.py into a common place (here?) than duplicate it.

Reply:
I'm here trying to get list of subvols/grp & for each path I parse I check if its present in this list. 
Ok so to explain why I'm doing this, currently I'm not sure if it's safe to assume that for a given path the position of subvol/grp will be the same (can we?) . For e.g. here
`/volumes/<fs_name>/<subvol grp name>/<subvol name>` 
can we assume that  `path.split[2]` or so will give us subvolume group name always? 
So I'm just applying brute-force way to check at each path parsing if it's a subvol/grp or not. does it makes sense?